### PR TITLE
Update steps for releasing new versions of plugins

### PR DESCRIPTION
**Pull request by shagunsodhani**
## Motivation

Update steps for releasing new versions of plugins

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebookresearch/hydra/blob/master/CONTRIBUTING.md)?

Yes/No

## Test Plan

(How should this PR be tested? Do you require special setup to run the test or repro the fixed bug?)

## Related Issues and PRs

(Is this PR part of a group of changes? Link the other relevant PRs and Issues here. Use https://help.github.com/en/articles/closing-issues-using-keywords for help on GitHub syntax)

### CODE REVIEW

File: website/docs/development/documentation.md [hunk 1]
```diff
@@ -25,6 +25,7 @@ issue number with one of the following extensions:
 * `docs` : Addition or updates to documentation
 * `config` : Changes or addition to the configuration structure
 * `maintenance` : Changes that improve the maintainability of the code
+* `trivial`: Minor Changes
```
Comment (omry) [inline website/docs/development/documentation.md:28]: I am actually not sure what purpose trivial news fragments are serving.

File: website/docs/development/release.md [hunk 2]
```diff
@@ -7,8 +7,8 @@ sidebar_label: Release process
 The release process may be automated in the future.
 
 - Checkout master
-- Update the Hydra version in `hydra/__init__.py`
-- Update NEWS.md with towncrier
-- Create a pip package for hydra-core: `python setup.py sdist bdist_wheel`
+- If releasing Hydra, update the version in `hydra/__init__.py`. For plugins, switch directory to `plugins/name-of-plugin` and update `setup.py`
+- Create draft release notes using `towncrier build --draft`. In case of plugins, add an extra flag for version i.e., `towncrier build --draft  --version x.y`
+- If the draft notes are correct, update NEWS.md using `towncrier build`. For plugins, use `towncrier build  --version x.y`
+- Create a pip package: `python setup.py sdist bdist_wheel`
```
Comment (omry) [inline website/docs/development/release.md:13]: Please use two different sections, one for Hydra and one for plugins.

For both plugins and Hydra, we should call-out that the release should be done from the Hydra release branch and not from master.

**Review by omry — CHANGES_REQUESTED** [review]

File: website/docs/development/documentation.md (hunk 1 above)
Comment (shagunsodhani) [reply website/docs/development/documentation.md:28]: `trivial` is already supported, just not documented. Should we remove the support for `trivial` ?

Comment (omry) [reply website/docs/development/documentation.md:28]: Yeah, let's just remove it.

File: website/docs/development/release.md [hunk 3]
```diff
@@ -6,9 +6,17 @@ sidebar_label: Release process
...
-- Update NEWS.md with towncrier
-- Create a pip package for hydra-core: `python setup.py sdist bdist_wheel`
+- Checkout the release branch.
+- Update the version in `hydra/__init__.py`.
+- Create draft release notes using `towncrier build --draft`.
+- If the draft notes are correct, update NEWS.md using `towncrier build`.
+- Create a pip package: `python setup.py sdist bdist_wheel`
+- Upload pip package: `python -m twine upload dist/*`
+
+
```
Comment (omry) [inline website/docs/development/release.md:16]: Section title?

**Review by omry — COMMENTED** [review]
The versioning strategy is that the repo is configured to be on the next version and the version should be bumped AFTER the release, not before.

One should check that the version is the correct one against pypi before releasing.
The tool tools/release/release.py does that by default, try it.
It can actually also build the wheels for each plugin. right now it's geared toward releasing many plugins at once but it can be tweaked to support a single plugin better.

File: website/docs/development/release.md [hunk 4]
```diff
@@ -6,17 +6,26 @@ sidebar_label: Release process
...
 
+The versioning strategy is that the repo is configured to be on the next version and the version should be bumped AFTER the release, not before.
+
+One should check that the version is the correct one against pypi before releasing.
+
+### Hydra Release Process
+
 - Checkout the release branch.
-- Update the version in `hydra/__init__.py`.
+- Check the version in `hydra/__init__.py`. This should be one version higher than the version available on PyPI. If it is the same version as PyPI, bump the version locally.
```
Comment (omry) [inline website/docs/development/release.md:16]: When do you commit the updated news file?

Comment (shagunsodhani) [reply website/docs/development/release.md:16]: Should be the last step, after bumping the version. Will update

Comment (shagunsodhani) [reply website/docs/development/release.md:16]: Added in https://github.com/facebookresearch/hydra/pull/1385/commits/e63c9a6f189696331b7a1c19e7ef8e1e4064699d

File: website/docs/development/release.md [hunk 5]
```diff
@@ -6,9 +6,28 @@ sidebar_label: Release process
...
+### Hydra Release Process
+
+- Checkout the release branch.
+- Check the version in `hydra/__init__.py`. This should be one version higher than the version available on PyPI. If it is the same version as PyPI, bump the version locally.
+- Create draft release notes using `towncrier build --draft`.
+- If the draft notes are correct, update NEWS.md using `towncrier build`.
+- Push the news file to release branch.
+- Create a pip package: `python setup.py sdist bdist_wheel`
+- Upload pip package: `python -m twine upload dist/*`
+- Bump the version in `hydra/__init__.py` and push to master.
```
Comment (omry) [inline website/docs/development/release.md:22]: You need to update the version in the release branch, not in master.

File: website/docs/development/release.md [hunk 6]
```diff
@@ -6,9 +6,28 @@ sidebar_label: Release process
...
+
+- Checkout the release branch and switch directory to `plugins/name-of-plugin`.
+- Bump the version in `setup.py`. This should be one version higher than the version available on PyPI. If it is the same version as PyPI, bump the version locally.
+- Create draft release notes using `towncrier build --draft  --version x.y`
+- If the draft notes are correct, update NEWS.md using `towncrier build  --version x.y`
+- Push the news file to release branch.
+- Create a pip package: `python setup.py sdist bdist_wheel`
 - Upload pip package: `python -m twine upload dist/*`
- 
+- Bump the version in `setup.py` and push to master.
```
Comment (omry) [inline website/docs/development/release.md:33]: Your instructions are bumping twice.
Please proof read.

File: website/docs/development/documentation.md (hunk 1 above)
Comment (shagunsodhani) [reply website/docs/development/documentation.md]: Done in https://github.com/facebookresearch/hydra/pull/1385/commits/eb893120149d00a66699c7c57f55d616dcb99818

File: website/docs/development/release.md (hunk 2 above)
Comment (shagunsodhani) [reply website/docs/development/release.md]: Done

File: website/docs/development/release.md (hunk 6 above)
Comment (shagunsodhani) [reply website/docs/development/release.md]: Fixed.

File: website/docs/development/release.md (hunk 5 above)
Comment (shagunsodhani) [reply website/docs/development/release.md]: Fixed.

File: website/docs/development/release.md [hunk 7]
```diff
@@ -6,9 +6,28 @@ sidebar_label: Release process
...
+
+One should check that the version is the correct one against pypi before releasing.
+
+### Hydra Release Process
+
+- Checkout the release branch.
+- Check the version in `hydra/__init__.py`. This should be one version higher than the version available on PyPI. If it is the same version as PyPI, bump the version locally.
+- Create draft release notes using `towncrier build --draft`.
+- If the draft notes are correct, update NEWS.md using `towncrier build`.
+- Push the news file to release branch.
```
Comment (omry) [inline website/docs/development/release.md:19]: Do not push anything directly.
Always create a PR, always wait for it's CI to complete successfully.

And wait for the CI to complete successfully.
```suggestion
- Push the news file to release branch.
```

File: website/docs/development/release.md [hunk 8]
```diff
@@ -6,9 +6,32 @@ sidebar_label: Release process
...
-- Update NEWS.md with towncrier
-- Create a pip package for hydra-core: `python setup.py sdist bdist_wheel`
+The versioning strategy is that the repo is configured to be on the next version and the version should be bumped AFTER the release, not before.
+
+One should check that the version is the correct one against pypi before releasing.
+
+### Hydra Release Process
+
+- Checkout the release branch.
+- Check the version in `hydra/__init__.py`. This should be one version higher than the version available on PyPI. If it is the same version as PyPI, bump the version locally.
```
Comment (jieru-hu) [inline website/docs/development/release.md:16]: nit: add link so life is easier :) 
[PyPI](https://pypi.org/project/hydra-core/)

File: website/docs/development/release.md [hunk 9]
```diff
@@ -6,9 +6,32 @@ sidebar_label: Release process
...
-- Update the Hydra version in `hydra/__init__.py`
-- Update NEWS.md with towncrier
-- Create a pip package for hydra-core: `python setup.py sdist bdist_wheel`
+The versioning strategy is that the repo is configured to be on the next version and the version should be bumped AFTER the release, not before.
+
+One should check that the version is the correct one against pypi before releasing.
+
+### Hydra Release Process
+
+- Checkout the release branch.
```
Comment (jieru-hu) [inline website/docs/development/release.md:15]: - Change branch to release branch and git pull to make sure on top of latest commit
- Checkout a local branch for releasing

File: website/docs/development/release.md [hunk 10]
```diff
@@ -6,9 +6,32 @@ sidebar_label: Release process
...
+- Check the version in `hydra/__init__.py`. This should be one version higher than the version available on PyPI. If it is the same version as PyPI, bump the version locally.
+- Create draft release notes using `towncrier build --draft`.
+- If the draft notes are correct, update NEWS.md using `towncrier build`.
```
Comment (jieru-hu) [inline website/docs/development/release.md:18]: Make sure the draft notes look correct,

File: website/docs/development/release.md [hunk 11]
```diff
@@ -6,9 +6,32 @@ sidebar_label: Release process
...
+- Push the news file to release branch.
```
Comment (jieru-hu) [inline website/docs/development/release.md:19]: Create a PR with the NEWS.md updated, wait for review and CI to pass.

File: website/docs/development/release.md [hunk 12]
```diff
@@ -6,9 +6,32 @@ sidebar_label: Release process
...
+- Create a pip package: `python setup.py sdist bdist_wheel`
+- Upload pip package: `python -m twine upload dist/*`
+- Bump the version in `hydra/__init__.py`.
```
Comment (jieru-hu) [inline website/docs/development/release.md:22]: (Before bumping the version)
* [Create a release](https://github.com/facebookresearch/hydra/releases/new) on GitHub
  - Tag the version, tags start with `v`, for example `v1.0.4`
  - Tag target to be the release branch
  - Set release title to be Hydra with release version, for example `Hydra 1.0.4`
  - Copy the release notes from added changes in NEWS.md from previous pull request. 
  - Publish the release

**Review by jieru-hu — CHANGES_REQUESTED** [review]
thank you for updating this! some comments below

File: website/docs/development/release.md [hunk 13]
```diff
@@ -6,9 +6,32 @@ sidebar_label: Release process
...
 
-- Checkout master
-- Update the Hydra version in `hydra/__init__.py`
-- Update NEWS.md with towncrier
-- Create a pip package for hydra-core: `python setup.py sdist bdist_wheel`
+The versioning strategy is that the repo is configured to be on the next version and the version should be bumped AFTER the release, not before.
+
+One should check that the version is the correct one against pypi before releasing.
+
+### Hydra Release Process
```
[the remark this answers is not in the document]
Comment (shagunsodhani) [reply website/docs/development/release.md:13]: Hmm. I do not have a strong preference. Will check with @omry to be double sure.

File: website/docs/development/release.md (hunk 12 above)
Comment (shagunsodhani) [reply website/docs/development/release.md:22]: Ohh good point!

File: plugins/hydra_nevergrad_sweeper/pyproject.toml [hunk 14]
```diff
@@ -33,11 +33,6 @@
   name = "Improved Documentation"
   showcontent = true
 
-  [[tool.towncrier.type]]
```
[the remark this answers is not in the document]
Comment (shagunsodhani) [reply plugins/hydra_nevergrad_sweeper/pyproject.toml:36]: Not directly related to the PR. I dont mind reverting and adding a new PR

[the remark this answers is not in the document]
Comment (shagunsodhani) [reply plugins/hydra_nevergrad_sweeper/pyproject.toml]: https://github.com/facebookresearch/hydra/pull/1406

File: website/docs/development/documentation.md (hunk 1 above)
Comment (jieru-hu) [inline website/docs/development/documentation.md:28]: remove?

File: website/docs/development/release.md [hunk 15]
```diff
@@ -6,9 +6,43 @@ sidebar_label: Release process
...
+- Update NEWS.md using `towncrier build`.
+- Create a pull request (to the release branch) with the NEWS.md updated, wait for review and CI to pass.
+- Create a pip package: `python setup.py sdist bdist_wheel`
+- Upload pip package: `python -m twine upload dist/*`
+- [Create a release on GitHub](https://github.com/facebookresearch/hydra/releases/new)
+- Tag the version, tags start with v, for example v1.0.4
+- Tag target to be the release branch.
+- Set release title to be Hydra with release version, for example "Hydra 1.0.4"
+- Copy the release notes from added changes in NEWS.md from previous pull request.
+- Publish the release.
```
Comment (jieru-hu) [inline website/docs/development/release.md:29]: Can you make these substeps of `Create a release on GitHub`? :)

File: website/docs/development/release.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (omry) [reply website/docs/development/release.md:13]: Why not both?
```suggestion
### Hydra Release Process (hydra-core)
```

File: website/docs/development/release.md [hunk 16]
```diff
@@ -6,9 +6,43 @@ sidebar_label: Release process
...
+One should check that the version is the correct one against pypi before releasing.
+
+### Hydra Release Process
+
+- Change branch to release branch and git pull to make sure on top of latest commit.
+- Checkout a local branch for releasing.
+- Check the version in `hydra/__init__.py`. This should be one version higher than the version available on [PyPI](https://pypi.org/project/hydra-core/). If it is the same version as PyPI, bump the version locally.
+- Create draft release notes using `towncrier build --draft`.
+- Make sure the draft notes look correct.
+- Update NEWS.md using `towncrier build`.
```
Comment (omry) [inline website/docs/development/release.md:20]: I don't know what version of towncrier you are using, but 19.2.0 (latest) does not have a build command.

File: website/docs/development/release.md (hunk 12 above)
Comment (omry) [reply website/docs/development/release.md:30]: I think creating a GitHub release should be done last.

**Comment by omry** [comment]
ping.

**Comment by omry** [comment]
Closing.

**Pull request closed without merging**